### ISSUE TITLE
[v4] New feature: on-attach/on-detach (bonus feature!)

### ISSUE THREAD

**Issue by patrick-steele-idem** [marko-js/marko#569]
It would be nice to be able to animate elements in and out of the DOM. For example:

```xml
style {
  li.animated {
    transition: max-height 250ms ease-in-out;
  }
}

class {
    animateIn(event, node) {
        node.style.maxHeight = '0px';
        setTimeout(() => {
            node.style.maxHeight = '250px';
        }, 0);
    }

    animateOut(event, node) {
        // Prevent the DOM node from being removed from the DOM so that
        // we can animate it out...
        event.preventDefault();
        
        node.style.maxHeight = '250px';
        setTimeout(() => {
            node.style.maxHeight = '0px';
            setTimeout(() => {
                // Animation has completed. Now we can animate it out
                event.detach();
            }, 250)
        }, 0);
    }
}
<div>
    <ul>
        <for(color in input.colors)>
            <li.animated ref=color on-attach('animateIn') on-detach('animateOut')>
                ${color}
            </li>
        </for>
    </ul>
</div>
```
